python/lolopy/learners.py

```diff
@@ -0,0 +1,204 @@
+from abc import abstractmethod, ABCMeta
+
+import numpy as np
+from lolopy.loloserver import get_java_gateway
+from py4j.java_collections import ListConverter
+from sklearn.base import BaseEstimator, RegressorMixin, ClassifierMixin, is_regressor
+
+
+class BaseLoloLearner(BaseEstimator, metaclass=ABCMeta):
+    """Base object for all leaners that use Lolo.
+
+    Contains logic for starting the JVM gateway, and the fit operations.
+    It is only necessary to implement the `_make_learner` object for """
+
+    def __init__(self):
+        self.gateway = get_java_gateway()
+
+        # Create a placeholder for the model
+        self.model_ = None
+
+    def fit(self, X, y, weights=None):
+
+        # Instantiate the JVM object
+        learner = self._make_learner()
+
+        # Convert all of the training data to Java arrays
+        train_data, weights_java = self._convert_train_data(X, y, weights)
+
+        # Make the weights
+        if weights is None:
+            result = learner.train(train_data)
+        else:
+            # Call training with weights
+            #  We encapsulate weights in "scale.Some" to make it an "Optional" class
+            result = learner.train(train_data, self.gateway.jvm.scala.Some(weights_java))
+
+        # Get the model out
+        self.model_ = result.getModel()
+
+        return self
+
+    @abstractmethod
+    def _make_learner(self):
+        """Instantiate the learner used by Lolo to train a model
+
+        Ret"""
+        pass
+
+    def _convert_train_data(self, X, y, weights=None):
+        """Convert the training data to a form accepted by Lolo
+
+        Args:
+            X (ndarray): Input variables
+            y (ndarray): Output variables
+            weights (ndarray): Wegihts for each sample
+        Returns
+            train_data (JavaObject): Pointer to the training data in Java
+            weights_java (JavaObject): POinter to the weights in Java, if provided
+        """
+
+        # Convert X and y to Java Objects
+        train_data = self.gateway.jvm.java.util.ArrayList(len(y))
+        for x_i, y_i in zip(X, y):
+            # Copy the X into an array
+            x_i_java = ListConverter().convert(x_i, self.gateway._gateway_client)
+            x_i_java = self.gateway.jvm.scala.collection.JavaConverters.asScalaBuffer(
+                x_i_java).toVector()
+
+            # If a regression problem, make sure y_i is a float
+            if is_regressor(self):
+                y_i = float(y_i)
+            else:
+                y_i = int(y_i)
```

Comment:
Does this imply that string labels are not supported?

Reply:
That is correct. I need the class labels to have a defined ordering, and the easiest route to that is to use integers and employ the [LabelEncoder](https://scikit-learn.org/stable/modules/generated/sklearn.preprocessing.LabelEncoder.html) built in to ScikitLearn for strings, etc.